File: routine/changes.py
"""What changed between two standings tables, as update-log entries.

The routine runs this instead of comparing the tables by eye: that is how the
Orioles' elimination went unlogged. It reads five JSON files and prints a
JSON list of log entries WITHOUT `at`; the routine adds that and appends
every entry, in order, in the same write as the change itself.

  python3 changes.py project NEW_STANDINGS
  python3 changes.py OLD_STANDINGS NEW_STANDINGS OLD_TEAMS NEW_TEAMS GAMES LOG BASELINE

The first form prints the projected field (`teams`) from the table you built,
so the projection is computed, not judged. The second prints the log entries.

OLD_STANDINGS / NEW_STANDINGS: the standings document as read and as built
({divisions: {...}}). OLD_TEAMS / NEW_TEAMS: the season document's `teams`
as read and as it will be written. Pass the same file twice for the teams
when the field did not change. GAMES: the `games` list the routine writes
to `slate.today` (away, home, state, score), from which each entry gets its
`via` -- the finals behind it -- so the log says WHY: "Orioles eliminated —
White Sox beat the Royals 9-1". LOG: the season document's `log` as read.
BASELINE: the `routine/baseline` document -- what was already clinched or
eliminated when the log began.

Comparing the two tables catches what changed on this run. The last part,
the CATCH-UP, catches what an earlier run missed, whatever the reason: every
club MLB currently marks as eliminated or clinched must have a log entry (or
be in the baseline), and any that doesn't gets one now, marked `late`. That
is the check that would have caught the White Sox, the Blue Jays and the
wild card clinches that went unlogged for days.
"""
import json
import sys


def rows(doc):
    """Every club's standings row, by id, with its division."""
    out = {}
    for div, clubs in (doc.get("divisions") or {}).items():
        for r in clubs:
            out[r["id"]] = dict(r, div=div)
    return out


def out_of_it(r):
    return r.get("elim") == "E" and r.get("wce") == "E"


def games(v):
    try:
        return float(str(v).lstrip("+"))
    except ValueError:
        return 0.0  # "-": leading, or level with the spot


def best_back(r):
    """Games back on the club's best remaining route: its division, or the
    wild card, whichever it is still alive for and closer in."""
    routes = []
    if r.get("elim") != "E":
        routes.append(games(r.get("gb")))
    if r.get("wce") != "E":
        routes.append(games(r.get("wcgb")))
    return f"{min(routes):.1f}" if routes else None


def result(games, club):
    """A club's final today as a `via` item, own runs first, or None."""
    for g in games:
        if g.get("state") != "final" or club not in (g["away"], g["home"]):
            continue
        a, h = g["score"]
        own, opp_runs, opp = (a, h, g["home"]) if g["away"] == club else (h, a, g["away"])
        return {"team": club, "won": own > opp_runs, "opp": opp, "score": [own, opp_runs]}
    return None


def via(*items):
    return [v for v in items if v]


def last_wild_card(after, league_of, lg):
    """The club holding the league's last wild card spot."""
    holders = [r for i, r in after.items()
               if league_of(i) == lg and r.get("wcrank") and not r.get("lead")]
    holders.sort(key=lambda r: int(r["wcrank"]))
    return holders[2]["id"] if len(holders) >= 3 else None


CLINCH_WHAT = {"x": "playoff", "w": "wildcard", "y": "division", "z": "bye"}
CLINCH_STEP = {"x": 1, "w": 2, "y": 3, "z": 4}
STEP_OF_WHAT = {w: CLINCH_STEP[c] for c, w in CLINCH_WHAT.items()}


def catch_up(after, entries, log, baseline):
    """Entries for anything MLB shows that neither the log, the baseline, nor
    this run's entries account for."""
    seen = list(log) + list(entries)
    out_logged = set(baseline.get("out", [])) | {e.get("team") for e in seen if e.get("kind") == "elim"}
    step = {i: CLINCH_STEP.get(c, 0) for i, c in (baseline.get("clinch") or {}).items()}
    for e in seen:
        if e.get("kind") == "berth":
            step[e["team"]] = max(step.get(e["team"], 0), STEP_OF_WHAT.get(e.get("what"), 0))
    late = []
    for i, r in after.items():
        c = r.get("clinch")
        if c in CLINCH_STEP and CLINCH_STEP[c] > step.get(i, 0):
            e = {"kind": "berth", "team": i, "what": CLINCH_WHAT[c], "late": True}
            if e["what"] == "division":
                e["div"] = r["div"]
            late.append(e)
    for i, r in after.items():
        if out_of_it(r) and i not in out_logged:
            late.append({"kind": "elim", "team": i, "late": True})
    return late


def project(st):
    """The field if the season ended today, per league: the three division
    leaders seeded 1-3 by win percentage, then the top three by wild card
    rank seeded 4-6. Ties keep MLB's own order (the table is in it)."""
    teams = {}
    for lg in ("AL", "NL"):
        divs = [d for d in st["divisions"] if d.startswith(lg)]
        leaders = [next(r for r in st["divisions"][d] if r.get("lead")) for d in divs]
        leaders.sort(key=lambda r: -float(r["pct"]))
        wild = sorted((r for d in divs for r in st["divisions"][d] if r.get("wcrank") and not r.get("lead")),
                      key=lambda r: int(r["wcrank"]))[:3]
        for seed, r in enumerate(leaders + wild, 1):
            teams[r["id"]] = {"league": lg, "seed": seed, "w": r["w"], "l": r["l"]}
    return teams


def main(old_st, new_st, old_teams, new_teams, games, log=None, baseline=None):
    before, after = rows(old_st), rows(new_st)
    entries = []
    league_of = lambda i: after[i]["div"][:2] if i in after else ""

    # The field: who came in, who went out, per league, paired in seed order.
    for lg in ("AL", "NL"):
        ins = sorted((i for i, t in new_teams.items() if t["league"] == lg and i not in old_teams),
                     key=lambda i: new_teams[i]["seed"])
        outs = sorted((i for i, t in old_teams.items() if t["league"] == lg and i not in new_teams),
                      key=lambda i: old_teams[i]["seed"])
        for i, o in zip(ins, outs):
            e = {"kind": "field", "in": i, "out": o}
            v = via(result(games, i), result(games, o))
            if v:
                e["via"] = v
            if new_teams[i]["seed"] <= 3:
                e.update(spot="division", div=after.get(i, {}).get("div", ""))
            else:
                e["spot"] = "wildcard"
            r = after.get(o)
            if r:
                e["outAlive"] = not out_of_it(r)
                back = best_back(r)
                if back is not None:
                    e["outBack"] = back
            entries.append(e)
        for i in ins[len(outs):]:
            entries.append({"kind": "field", "in": i})
        for o in outs[len(ins):]:
            entries.append({"kind": "field", "out": o})

    # Seed moves among clubs that stayed in the field: only moves UP, since
    # every move up implies one down. `over` names the club passed when the
    # two simply swapped seeds.
    for lg in ("AL", "NL"):
        stayed = [i for i, t in new_teams.items() if t["league"] == lg and i in old_teams]
        for i in sorted(stayed, key=lambda i: new_teams[i]["seed"]):
            was, now = old_teams[i]["seed"], new_teams[i]["seed"]
            if now >= was:
                continue
            e = {"kind": "seed", "team": i, "from": was, "to": now}
            swap = [j for j in stayed if old_teams[j]["seed"] == now and new_teams[j]["seed"] == was]
            if swap:
                e["over"] = swap[0]
            v = via(result(games, i), result(games, e["over"]) if swap else None)
            if v:
                e["via"] = v
            entries.append(e)

    # Clinches on this run, from MLB's clinchIndicator (`clinch` in the table):
    # x a playoff spot, w a wild card, y the division, z a first-round bye.
    # Each step up is its own news -- the White Sox clinching a spot and,
    # later, a wild card are two entries. A table read before `clinch` was
    # recorded has nothing to compare against, so only division titles (from
    # `clinched`) can be found then.
    WHAT, STEP = CLINCH_WHAT, CLINCH_STEP
    for i, r in after.items():
        old = before.get(i, {})
        if "clinch" in old:
            new_c, old_c = r.get("clinch"), old.get("clinch")
            if new_c not in WHAT or STEP[new_c] <= STEP.get(old_c, 0):
                continue
            e = {"kind": "berth", "team": i, "what": WHAT[new_c]}
        elif r.get("clinched") and not old.get("clinched"):
            e = {"kind": "berth", "team": i, "what": "division"}
        else:
            continue
        if e["what"] == "division":
            e["div"] = r["div"]
        own = result(games, i)
        if own and own["won"]:
            e["via"] = [own]
        entries.append(e)

    # Clubs whose season ended on this run: out of the division AND the wild card.
    # Why: its own loss, and the win by the club it was chasing on the route
    # that closed this run -- the division leader, or the last wild card.
    for i, r in after.items():
        if out_of_it(r) and not out_of_it(before.get(i, {})):
            e = {"kind": "elim", "team": i}
            old = before.get(i, {})
            chasing = None
            if old.get("wce") != "E":
                chasing = last_wild_card(after, league_of, league_of(i))
            elif old.get("elim") != "E":
                chasing = next((x for x, rr in after.items() if rr["div"] == r["div"] and rr.get("lead")), None)
            own = result(games, i)
            them = result(games, chasing) if chasing else None
            v = via(own if own and not own["won"] else None, them if them and them["won"] else None)
            if v:
                e["via"] = v
            entries.append(e)

    if log is not None:
        entries += catch_up(after, entries, log, baseline or {})
    return entries


if __name__ == "__main__":
    if sys.argv[1] == "project":
        # python3 changes.py project NEW_STANDINGS: the projected `teams`.
        print(json.dumps(project(json.load(open(sys.argv[2]))), indent=1))
        sys.exit()
    docs = [json.load(open(p)) for p in sys.argv[1:8]]
    if len(docs) > 5:
        # The season document as read: its log. The baseline document as read.
        docs[5] = docs[5].get("log", docs[5]) if isinstance(docs[5], dict) else docs[5]
    print(json.dumps(main(*docs), indent=1))

File: page/js/changes.js
export const MAX_LOG = 50;

function rowsById(standings) {
  const out = {};
  for (const [div, clubs] of Object.entries((standings && standings.divisions) || {})) {
    for (const r of clubs) out[r.id] = { ...r, div };
  }
  return out;
}

const outOfIt = (r) => !!r && r.elim === "E" && r.wce === "E";
const gamesBack = (v) => {
  const n = parseFloat(String(v).replace("+", ""));
  return isNaN(n) ? 0 : n;
};

function bestBack(r) {
  const routes = [];
  if (r.elim !== "E") routes.push(gamesBack(r.gb));
  if (r.wce !== "E") routes.push(gamesBack(r.wcgb));
  return routes.length ? Math.min(...routes).toFixed(1) : null;
}

function result(games, club) {
  const g = games.find((x) => x.state === "final" && club && (x.away === club || x.home === club));
  if (!g) return null;
  const [a, h] = g.score;
  const [own, theirs, opp] = g.away === club ? [a, h, g.home] : [h, a, g.away];
  return { team: club, won: own > theirs, opp, score: [own, theirs], end: g.end };
}

// `at` is when the change was noticed, not when its game ended, so a change found after
// the last dismissal still shows as new.
function entry(fields, via, now) {
  const games = via.filter(Boolean);
  const e = { ...fields };
  if (games.length) e.via = games.map(({ end, ...v }) => v);
  e.at = new Date(now).toISOString().replace(".000Z", "Z");
  return e;
}

function lastWildCard(rows, lg) {
  const holders = Object.values(rows)
    .filter((r) => r.div.startsWith(lg) && r.wcrank && !r.lead)
    .sort((a, b) => Number(a.wcrank) - Number(b.wcrank));
  return holders.length >= 3 ? holders[2].id : null;
}

// Seed moves are logged only when the field is unchanged, since a swap moves seeds too,
// and only upward, since every rise implies a fall.
function fieldChanges(oldTeams, newTeams, after, games, now, logSeeds) {
  const out = [];
  for (const lg of ["AL", "NL"]) {
    const league = (teams) => Object.entries(teams).filter(([, t]) => t.league === lg);
    const bySeed = (teams, ids) => ids.sort((a, b) => teams[a].seed - teams[b].seed);
    const ins = bySeed(
      newTeams,
      league(newTeams)
        .map(([id]) => id)
        .filter((id) => !oldTeams[id]),
    );
    const outs = bySeed(
      oldTeams,
      league(oldTeams)
        .map(([id]) => id)
        .filter((id) => !newTeams[id]),
    );

    ins.forEach((id, i) => {
      const gone = outs[i];
      if (!gone) {
        out.push(entry({ kind: "field", in: id }, [], now));
        return;
      }
      const fields = { kind: "field", in: id, out: gone };
      if (newTeams[id].seed <= 3)
        Object.assign(fields, { spot: "division", div: (after[id] || {}).div || "" });
      else fields.spot = "wildcard";
      const r = after[gone];
      if (r) {
        fields.outAlive = !outOfIt(r);
        const back = bestBack(r);
        if (back != null) fields.outBack = back;
      }
      out.push(entry(fields, [result(games, id), result(games, gone)], now));
    });
    outs.slice(ins.length).forEach((id) => out.push(entry({ kind: "field", out: id }, [], now)));

    if (!logSeeds || ins.length || outs.length) continue;
    const moved = league(newTeams).filter(([id, t]) => t.seed < oldTeams[id].seed);
    for (const [id, t] of moved) {
      const fields = { kind: "seed", team: id, from: oldTeams[id].seed, to: t.seed };
      const passed = league(newTeams).filter(
        ([other, o]) =>
          other !== id && oldTeams[other].seed === t.seed && o.seed === oldTeams[id].seed,
      );
      const swapOnly = moved.length === 1 && passed.length === 1;
      if (swapOnly) fields.over = passed[0][0];
      out.push(
        entry(fields, [result(games, id), swapOnly ? result(games, fields.over) : null], now),
      );
    }
  }
  return out;
}

// Each step up MLB's clinch marker is its own news.
const CLINCH = { x: ["playoff", 1], w: ["wildcard", 2], y: ["division", 3], z: ["bye", 4] };

// A stored table without `clinch` can only reveal a division title.
function berthWon(old, r) {
  if ("clinch" in old) {
    const now = CLINCH[r.clinch],
      was = CLINCH[old.clinch];
    return now && now[1] > (was ? was[1] : 0) ? now[0] : null;
  }
  return r.clinched && !old.clinched ? "division" : null;
}

function standingsChanges(before, after, games, now) {
  const out = [];
  const clubs = Object.entries(after).filter(([id]) => before[id]);
  for (const [id, r] of clubs) {
    const what = berthWon(before[id], r);
    if (!what) continue;
    const fields = { kind: "berth", team: id, what };
    if (what === "division") fields.div = r.div;
    const own = result(games, id);
    out.push(entry(fields, [own && own.won ? own : null], now));
  }
  for (const [id, r] of clubs) {
    const old = before[id];
    if (!outOfIt(r) || outOfIt(old)) continue;
    const chasing =
      old.wce !== "E"
        ? lastWildCard(after, r.div.slice(0, 2))
        : old.elim !== "E"
          ? (Object.values(after).find((x) => x.div === r.div && x.lead) || {}).id
          : null;
    const own = result(games, id),
      them = result(games, chasing);
    out.push(
      entry(
        { kind: "elim", team: id },
        [own && !own.won ? own : null, them && them.won ? them : null],
        now,
      ),
    );
  }
  return out;
}

export function between(before, after, now = Date.now()) {
  if (!before || !after) return [];
  const games = (after.slate && after.slate.today && after.slate.today.games) || [];
  const oldTeams = before.teams || {},
    newTeams = after.teams || {};
  const out = [];

  const hadField = Object.keys(oldTeams).length > 0;
  const locked = before.projected !== false && after.projected === false;
  if (locked && hadField) out.push(entry({ kind: "lock" }, [], now));
  if (hadField && (after.projected !== false || locked)) {
    out.push(...fieldChanges(oldTeams, newTeams, rowsById(after.standings), games, now, !locked));
  }
  if (before.standings && before.standings.divisions && after.standings) {
    out.push(
      ...standingsChanges(rowsById(before.standings), rowsById(after.standings), games, now),
    );
  }
  return out;
}

// A field or seed change can recur on a later day, so its key carries the day.
export function key(e) {
  switch (e.kind) {
    case "game":
      return `game:${e.series}:${e.game}`;
    case "clinch":
      return `clinch:${e.series}`;
    case "lock":
      return "lock";
    case "elim":
      return `elim:${e.team}`;
    case "berth":
      return `berth:${e.team}:${e.what}`;
    case "field":
      return `field:${e.in || ""}:${e.out || ""}:${String(e.at).slice(0, 10)}`;
    case "seed":
      return `seed:${e.team}:${e.to}:${String(e.at).slice(0, 10)}`;
    default:
      return `${e.kind}:${e.at}:${e.text || ""}`;
  }
}

export function merge(log, entries) {
  const seen = new Set();
  const all = [...(log || []), ...entries].filter((e) => {
    if (!e || !e.kind) return false;
    const k = key(e);
    if (seen.has(k)) return false;
    seen.add(k);
    return true;
  });
  all.sort((a, b) => Date.parse(a.at) - Date.parse(b.at));
  return all.slice(-MAX_LOG);
}
